### PR TITLE
feat: add dry-run support for core deploy

### PR DESCRIPTION
**Pull request by nbayindirli**
**Closed in favor of https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/3558**

### Description

* New feature to allow users to dry-run the core deploy script against a forked network of their choice
* To run: `yarn build && yarn hyperlane deploy core --dry-run` || `yarn build && yarn hyperlane deploy core -d`
* Externally enables: `hyperlane deploy core --dry-run` || `hyperlane deploy core -d`

### Drive-by changes

* For all touched paths:
  * Code cleanliness + reuse
  * Method docs
  * Minor changes for increased symmetry

### Related issues

* [Issue 819](https://github.com/hyperlane-xyz/issues/issues/819)

### Backward compatibility

* Yes

### Testing

* Note: `hl` == `yarn build && yarn hyperlane`. The below tests are only a sample and are not inclusive.

#### Manual Testing
  * With `anvil` NOT running in separate instance:
    * `hl deploy core -d`
      * Throws:
```
Error: No active anvil node detected.
        Please run `anvil` in a separate instance.
```
  * With `anvil` running in separate instance:
     * `hl deploy core -d --targets alfajores -k c0052e22df5d1f4ae7c51e254Xx00Xx0eb833453eaed6301xXxxx8a30d92d10a` (any private key)
       * Throws `Error: Invalid address length. Please ensure you are passing an address and not a private key.`
     * `hl deploy core -d --targets bsctestnet -k 0x16F4898F47c085C41d7Cc6b1dc0xX0xXX017dcBb` (any public address; BSC public RPC endpoint does not support forking)
     * Output: `⛔️ [dry-run] The current RPC may not support forking. Please consider using a different RPC provider.`
     * `hl deploy core -d --targets alfajores -k 0x16F4898F47c085C41d7Cc6b1dc0xX0xXX017dcBb` (any public address)
     * Output:
```
Verifying anvil node is running...
Successfully verified anvil node is running ✅
No chain config file provided
Forking alfajores for dry-run...
Successfully forked alfajores for dry-run ✅
Impersonating account (0x16F4898F47c085C41d7Cc6b1dc72B91EA617dcBb)...
Successfully impersonated account (0x16F4898F47c085C41d7Cc6b1dc72B91EA617dcBb) ✅
...
Core contracts deployed
Writing agent configs
Agent configs written
Deployment is complete!
Contract address artifacts are in artifacts/dry-run_core-deployment-2024-02-23-15-51-45.json
Agent configs are in artifacts/dry-run_agent-config-2024-02-23-15-51-45.json
Resetting forked network...
Successfully reset forked network ✅
```

#### CI Testing
  * Successful CI-backed integration/regression testing via `ci-test.sh` (includes script refactor)

### CODE REVIEW

File: typescript/cli/anvil-start.sh
Comment (yorhodes) [inline typescript/cli/anvil-start.sh:1]: I would add an alias in the published CLI that launches the user-installed anvil/hardhat node binary instead of a shellscript like this which requires someone to build the repo

File: typescript/cli/src/commands/deploy.ts [hunk 1]
```diff
@@ -39,17 +50,11 @@ const agentCommand: CommandModule = {
...
-        type: 'string',
-        description: 'Comma separated list of chains to relay between',
-      },
+      origin: originCommandOption,
+      targets: agentTargetsCommandOption,
       chains: chainsCommandOption,
-      config: agentConfigurationOption,
-    }),
+      config: agentConfigCommandOption,
+    } as AgentCommandOptions),
```
Comment (yorhodes) [inline typescript/cli/src/commands/deploy.ts:57]: nit: replace this cast with a type assertion?

File: typescript/cli/src/commands/deploy.ts [hunk 2]
```diff
@@ -75,30 +80,17 @@ const coreCommand: CommandModule = {
...
-          'A path to a JSON or YAML file with Hook configs (for every chain)',
-      },
+      ism: ismCommandOption,
+      hook: hookCommandOption,
       out: outDirCommandOption,
       key: keyCommandOption,
       yes: skipConfirmationOption,
-    }),
+      'dry-run': dryRunOption,
+    } as CoreCommandOptions),
```
Comment (yorhodes) [inline typescript/cli/src/commands/deploy.ts:92]: ditto

File: typescript/cli/src/commands/deploy.ts [hunk 3]
```diff
@@ -131,17 +139,13 @@ const warpCommand: CommandModule = {
...
-        default: './configs/warp-tokens.yaml',
-      },
+      config: warpConfigCommandOption,
       core: coreArtifactsOption,
       chains: chainsCommandOption,
       out: outDirCommandOption,
       key: keyCommandOption,
       yes: skipConfirmationOption,
-    }),
+    } as WarpCommandOptions),
```
Comment (yorhodes) [inline typescript/cli/src/commands/deploy.ts:148]: ditto

File: typescript/cli/src/commands/options.ts [hunk 4]
```diff
@@ -1,10 +1,96 @@
...
+  yes: boolean;
+  'dry-run': boolean;
+}
+export type WarpCommandArgs = CommandArgs & {
+  config: string;
+  core: string;
+  out: string;
+  key: string;
+  yes: boolean;
+};
```
Comment (yorhodes) [inline typescript/cli/src/commands/options.ts:54]: is this all unused?

File: typescript/cli/src/commands/options.ts [hunk 5]
```diff
@@ -1,10 +1,96 @@
...
+export const warpConfigCommandOption: Options = {
+  type: 'string',
+  description: 'A path to a JSON or YAML file with a warp config.',
+  default: './configs/warp-tokens.yaml',
+};
+
+export const keyCommandOption: Options = {
+  type: 'string',
+  description: `Default: A hex private key or seed phrase for transaction signing, or use the HYP_KEY env var.
+Dry-run: A hex public key to simulate transaction signing on a forked network, or use the HYP_KEY env var.`,
```
Comment (yorhodes) [inline typescript/cli/src/commands/options.ts:93]: should we colocate this or share with env schema somehow?

File: typescript/cli/src/context.ts [hunk 6]
```diff
@@ -103,25 +92,25 @@ export async function getContext<P extends ContextSettings>({
...
+/**
+ * Retrieves a new MultiProvider based on all known chain metadata & custom user chains
+ * @param customChains Custom chains specified by the user
+ * @returns a new MultiProvider
+ */
+export function getMultiProvider(customChains: ChainMap<ChainMetadata>) {
   const chainConfigs = { ...chainMetadata, ...customChains };
-  const mp = new MultiProvider(chainConfigs);
-  if (signer) mp.setSharedSigner(signer);
-  return mp;
```
Comment (yorhodes) [inline typescript/cli/src/context.ts:126]: why remove optional signer here?

File: typescript/cli/src/utils/keys.ts [hunk 7]
```diff
@@ -1,18 +1,75 @@
...
+function privateKeyToSigner(key: string): ethers.Wallet {
+  if (!key) throw new Error('No private key provided');
+
+  const formattedKey = key.trim().toLowerCase();
+  if (ethers.utils.isHexString(ensure0x(formattedKey)))
+    return new ethers.Wallet(ensure0x(formattedKey));
+  else if (formattedKey.split(' ').length >= 6)
+    return ethers.Wallet.fromMnemonic(formattedKey);
+  else throw new Error('Invalid private key format');
+}
```
Comment (yorhodes) [inline typescript/cli/src/utils/keys.ts:75]: why move this and remove export?

File: typescript/cli/src/utils/fork.ts [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+ * @param account the EOAs public-key to impersonate
+ * @returns the impersonated signer
+ */
+export const impersonateAccount = async (
+  account: string,
+): Promise<providers.JsonRpcSigner> => {
+  const provider = getLocalProvider();
+  await provider.send(ANVIL_RPC_METHODS.IMPERSONATE_ACCOUNT, [account]);
+
+  return provider.getSigner(account);
```
Comment (yorhodes) [inline typescript/cli/src/utils/fork.ts:74]: I believe this endpoint works with an address, not the public key

File: typescript/cli/src/utils/keys.ts [hunk 9]
```diff
@@ -1,18 +1,75 @@
...
+ */
+async function publicKeyToSigner(
+  key: Address,
+): Promise<providers.JsonRpcSigner> {
+  if (!key) throw new Error('No public key provided');
+
+  const formattedKey = key.trim().toLowerCase();
+  if (ethers.utils.isHexString(ensure0x(formattedKey)))
+    return await impersonateAccount(key);
+  else throw new Error('Invalid public key format');
```
Comment (yorhodes) [inline typescript/cli/src/utils/keys.ts:58]: ditto on this being address, not public key
also I might rename this to make it more obvious its a mocked signer/only useful for testing

File: typescript/cli/src/utils/keys.ts [hunk 10]
```diff
@@ -1,18 +1,75 @@
...
+ * @returns the context-specific signer
+ */
+export async function getSigner<P extends ContextSettings>({
+  keyConfig,
+  skipConfirmation,
+  dryRun,
+}: P): Promise<providers.JsonRpcSigner | ethers.Wallet | undefined> {
+  if (!keyConfig) return undefined;
+
+  const keyType = dryRun ? 'public' : 'private';
```
Comment (yorhodes) [inline typescript/cli/src/utils/keys.ts:22]: ditto on this being address

File: typescript/cli/src/utils/keys.ts [hunk 11]
```diff
@@ -1,18 +1,75 @@
...
+  let key: string;
+  if (keyConfig.key) key = keyConfig.key;
+  else if (skipConfirmation) throw new Error(`No ${keyType} key provided`);
+  else
+    key = await input({
+      message:
+        keyConfig.promptMessage ||
+        `Please enter a ${keyType} key or use the HYP_KEY environment variable.`,
```
Comment (yorhodes) [inline typescript/cli/src/utils/keys.ts:30]: this also means we are overloading HYP_KEY with addresses and private keys?

File: typescript/cli/src/utils/fork.ts
Comment (yorhodes) [inline typescript/cli/src/utils/fork.ts:1]: flagging that viem has strong types around the hardhat RPCs https://v1.viem.sh/docs/actions/test/impersonateAccount.html#impersonateaccount

File: typescript/cli/src/utils/fork.ts [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+};
+
+/**
+ * Retrives a local provider. Defaults to DEFAULT_ANVIL_ENDPOINT.
```
Comment (yorhodes) [inline typescript/cli/src/utils/fork.ts:78]: nit: retrieves

File: typescript/cli/src/utils/fork.ts [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+    logYellow(
+      `⚠️ Provided URL override (${urlOverride}) does not begin with ${ENDPOINT_PREFIX}. Defaulting to ${
+        envUrl ?? DEFAULT_ANVIL_ENDPOINT
+      }`,
+    );
+    urlOverride = undefined;
+  }
+
+  return new providers.JsonRpcProvider(
+    urlOverride ?? envUrl ?? DEFAULT_ANVIL_ENDPOINT,
```
Comment (yorhodes) [inline typescript/cli/src/utils/fork.ts:99]: slightly hard to follow this
can we have only one variable here?

File: typescript/cli/src/utils/keys.ts [hunk 14]
```diff
@@ -1,18 +1,75 @@
...
+
+/**
+ * Generates a signer from a public key.
+ * @param key a public key
+ * @returns a signer for the public key
+ */
+async function publicKeyToSigner(
+  key: Address,
+): Promise<providers.JsonRpcSigner> {
+  if (!key) throw new Error('No public key provided');
```
Comment (yorhodes) [inline typescript/cli/src/utils/keys.ts:53]: imo zod is responsible for this type of assertion but I see that we already had this pattern

File: typescript/cli/src/utils/fork.ts [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+  chain: ChainName | number,
+) => {
+  logGray(`Forking ${chain} for dry-run...`);
+
+  const provider = getLocalProvider();
+  const currentChainMetadata = multiProvider.metadata[chain];
+
+  await provider.send(ANVIL_RPC_METHODS.SET_CHAIN_ID, [
+    currentChainMetadata.chainId,
+  ]);
```
Comment (yorhodes) [inline typescript/cli/src/utils/fork.ts:48]: do we even need to do this? I wasn't doing this in the other fork scripts iirc

File: typescript/cli/src/deploy/dry-run.ts [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+/**
+ * Evaluates if an error is related to the current dry-run.
+ * @param error the thrown error
+ * @param dryRun whether or not the current command is being dry-run
+ */
+export function evaluateIfDryRunFailure(error: any, dryRun: boolean) {
+  if (dryRun && error.message.includes('underlying network changed'))
+    logYellow(
+      "⚠️ (Dry-run) The network's selected RPC provider may not support forking. Please try running with another RPC provider.",
+    );
```
Comment (yorhodes) [inline typescript/cli/src/deploy/dry-run.ts]: im not convinced the "underlying network changed" error is related to the RPC not supporting forking but I could be wrong

File: typescript/cli/src/deploy/dry-run.ts [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+ * Ensures an anvil node is running locally.
+ */
+export async function verifyAnvil() {
+  const provider = getLocalProvider();
+  try {
+    await provider.send(ANVIL_RPC_METHODS.NODE_INFO, []);
+  } catch (error: any) {
+    if (error.message.includes('missing response'))
+      throw new Error(`No active anvil node detected.
+\tRun \`./hyperlane-monorepo/typescript/cli/anvil-start.sh\` or \`anvil\` in a separate instance to dry-run.`);
```
Comment (yorhodes) [inline typescript/cli/src/deploy/dry-run.ts:34]: ditto on this
lets alias a CLI comand for the latter? I dont think users should be in the monorepo

File: typescript/cli/src/deploy/dry-run.ts [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+  chains: string[],
+) {
+  if (chains.length != 1)
+    throw new Error('Only one chain can be present for dry-running.');
+  const chain = chains[0];
+  multiProvider = multiProvider.extendChainMetadata({
+    [chain]: { blocks: { confirmations: 0 } },
+  });
+  await setFork(multiProvider, chain);
+}
```
Comment (yorhodes) [inline typescript/cli/src/deploy/dry-run.ts:22]: if this is only dry run lets make the chain param only one?

File: typescript/cli/src/deploy/core.ts [hunk 19]
```diff
@@ -78,21 +89,30 @@ export async function runCoreDeploy({
...
+
+  const signer = (await getSigner({
+    keyConfig: { key },
+    skipConfirmation,
+    dryRun,
+  })) as ethers.Signer;
+
+  if (signer) multiProvider.setSharedSigner(signer);
+
+  if (dryRun) await prepareMultiProviderForDryRun(multiProvider, chains);
```
Comment (yorhodes) [inline typescript/cli/src/deploy/core.ts:114]: still seems like we are forking after setting the signer?
also this function isnt really "dryrun" specific so maybe rename

File: typescript/cli/src/deploy/core.ts [hunk 20]
```diff
@@ -350,6 +399,24 @@ async function executeDeploy({
...
+  const coreDeploymentFile = {
+    filename: dryRun ? 'dry-run_core-deployment' : 'core-deployment',
+    description: 'Contract addresses',
+  };
+  const agentConfigFile = {
+    filename: dryRun ? 'dry-run_agent-config' : 'agent-config',
+    description: 'Agent configs',
+  };
+
+  return [coreDeploymentFile, agentConfigFile];
```
Comment (yorhodes) [inline typescript/cli/src/deploy/core.ts:417]: for dry runs im not sure how useful the artifacts are?

File: typescript/cli/src/utils/fork.ts [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+const DEFAULT_ANVIL_ENDPOINT = 'http://127.0.0.1:8545';
+
+export enum ETH_RPC_METHODS {
+  BLOCK_NUMBER = 'eth_blockNumber',
+}
+
+export enum ANVIL_RPC_METHODS {
+  RESET = 'anvil_reset',
+  SET_RPC_URL = 'anvil_setRpcUrl',
+  SET_CHAIN_ID = 'anvil_setChainId',
```
Comment (yorhodes) [inline typescript/cli/src/utils/fork.ts:19]: I dont see either of these being used

File: typescript/cli/src/utils/fork.ts [hunk 22]
```diff
@@ -0,0 +1,101 @@
...
+import { logGray, logGreen, logYellow } from '../../logger.js';
+
+import { ENV } from './env.js';
+
+const ENDPOINT_PREFIX = 'http://';
+const DEFAULT_ANVIL_ENDPOINT = 'http://127.0.0.1:8545';
+
+export enum ETH_RPC_METHODS {
+  BLOCK_NUMBER = 'eth_blockNumber',
+}
```
Comment (yorhodes) [inline typescript/cli/src/utils/fork.ts:14]: is this unused? I assume we are just using the ethers RPC client everywhere

**Review by yorhodes — COMMENTED** [review]
looks like a good start

File: typescript/cli/src/commands/options.ts (hunk 4 above)
Comment (nbayindirli) [reply typescript/cli/src/commands/options.ts]: Thought I might use this in next iteration, but decided to remove. Removed now

File: typescript/cli/src/deploy/core.ts (hunk 19 above)
Comment (nbayindirli) [reply typescript/cli/src/deploy/core.ts]: Discussed offline– refactored control flow here and updated naming convention

File: typescript/cli/src/deploy/core.ts (hunk 20 above)
Comment (nbayindirli) [reply typescript/cli/src/deploy/core.ts:425]: This allows a developer to have evidence of past runs to see any diff in their deployed changes. Also, we may eventually want to use `anvil_dumpState` and `anvil_loadState` to allow developers to reload actual anvil runs to use the same addresses for dry-runs. See [here](https://book.getfoundry.sh/reference/anvil/#custom-methods) for those custom methods

File: typescript/cli/src/deploy/dry-run.ts (hunk 18 above)
Comment (nbayindirli) [reply typescript/cli/src/deploy/dry-run.ts]: Agreed. Done.

File: typescript/cli/src/deploy/dry-run.ts (hunk 17 above)
Comment (nbayindirli) [reply typescript/cli/src/deploy/dry-run.ts]: Also totally agreed. Boiled verbiage down to just running anvil standalone

File: typescript/cli/src/utils/fork.ts (hunk 22 above)
Comment (nbayindirli) [reply typescript/cli/src/utils/fork.ts]: Yep, this was within the scope of my cleanup post-wip. Now removed

File: typescript/cli/src/utils/fork.ts (hunk 21 above)
Comment (nbayindirli) [reply typescript/cli/src/utils/fork.ts]: Same as above

File: typescript/cli/src/utils/fork.ts (hunk 15 above)
Comment (nbayindirli) [reply typescript/cli/src/utils/fork.ts]: No we do not, and this is also post-wip cleanup. Now removed

File: typescript/cli/src/utils/keys.ts (hunk 7 above)
Comment (nbayindirli) [reply typescript/cli/src/utils/keys.ts:81]: It's used nowhere beyond `getSigner()` (above), so limited scope should always be preferred until necessary

File: typescript/cli/src/deploy/dry-run.ts (hunk 16 above)
Comment (nbayindirli) [reply typescript/cli/src/deploy/dry-run.ts]: You are correct– this is now updated accordingly

File: typescript/cli/anvil-start.sh [hunk 23]
```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/env bash
+
+# NOTE: This script is intended to be run from the root of the repo
+
+echo -e "\nYou must have 'anvil' from Foundry installed to proceed."
+echo -e '\nTo install, run the following command:'
+echo -e '\tcurl -L https://foundry.paradigm.xyz | bash && foundryup && anvil\n'
```
Comment (jmrossy) [inline typescript/cli/anvil-start.sh:7]: Lets point people to foundry's docs instead of specifying instructions here. The steps may change or vary based on platform.

File: typescript/cli/anvil-start.sh [hunk 24]
```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/env bash
+
+# NOTE: This script is intended to be run from the root of the repo
```
Comment (jmrossy) [inline typescript/cli/anvil-start.sh:3]: This is true of the CLI CI test. Is it also required for this?

File: typescript/cli/src/commands/deploy.ts [hunk 25]
```diff
@@ -109,16 +101,32 @@ const coreCommand: CommandModule = {
...
+        ismConfigPath,
+        hookConfigPath,
+        outPath,
+        skipConfirmation,
+        dryRun,
+      });
+    } catch (error: any) {
+      evaluateIfDryRunFailure(error, dryRun);
+
+      throw new Error(error);
```
Comment (jmrossy) [inline typescript/cli/src/commands/deploy.ts:128]: You don't need to re-wrap the error here, you can just `throw error`

File: typescript/cli/src/utils/fork.ts (hunk 13 above)
Comment (yorhodes) [reply typescript/cli/src/utils/fork.ts]: why not use the same variable for envUrl and urlOverride too?

File: typescript/cli/src/commands/deploy.ts [hunk 26]
```diff
@@ -130,12 +138,8 @@ const warpCommand: CommandModule = {
...
   describe: 'Deploy Warp Route contracts',
   builder: (yargs) =>
-    yargs.options({
-      config: {
-        type: 'string',
-        description: 'A path to a JSON or YAML file with a warp config.',
-        default: './configs/warp-tokens.yaml',
-      },
+    yargs.options<WarpCommandOptions>({
+      config: warpConfigCommandOption,
```
Comment (jmrossy) [inline typescript/cli/src/commands/deploy.ts:142]: I'm not totally convinced it's worth moving options to the shared options file if they're only used on one place but ndb. It does look neater this way.

File: typescript/cli/src/context.ts [hunk 27]
```diff
@@ -45,20 +46,24 @@ export function getMergedContractAddresses(
...
-  keyConfig?: {
-    key?: string;
-    promptMessage?: string;
-  };
+  keyConfig?: KeyConfig;
   skipConfirmation?: boolean;
 }
 
 interface CommandContextBase {
+  chains: string[];
```
Comment (jmrossy) [inline typescript/cli/src/context.ts:66]: Nit: `ChainName[]`

File: typescript/cli/src/context.ts [hunk 28]
```diff
@@ -116,12 +116,71 @@ export async function getContext<P extends ContextSettings>({
...
+        message:
+          coreConfig.promptMessage ||
+          'Do you want to use some core deployment address artifacts? This is required for PI chains (non-core chains).',
+        skipConfirmation,
+      })) || {};
+  }
+
+  const multiProvider = getMultiProvider(customChains);
+
+  if (!chains?.length) {
```
Comment (jmrossy) [inline typescript/cli/src/context.ts:146]: What happens if multiple chains are provided via the param here?

File: typescript/cli/src/utils/fork.ts [hunk 29]
```diff
@@ -0,0 +1,131 @@
...
+
+/**
+ * Stops account impersonation.
+ * @param address the address to stop impersonating
+ */
+export const stopImpersonatingAccount = async (address: Address) => {
+  logGray(`Stopping account impersonation for address (${address})...`);
+
+  const trimmedAddress = address.substring(2);
+  if (trimmedAddress.length != TRIMMED_ETHEREUM_ADDRESS_LENGTH)
```
Comment (jmrossy) [inline typescript/cli/src/utils/fork.ts:92]: Why trim here instead of comparing to size 42?
Also, the functions in the `utils` package would be useful here (see `isValidAddress`)

File: typescript/cli/src/utils/keys.ts [hunk 30]
```diff
@@ -1,18 +1,109 @@
...
+import { ethers, providers } from 'ethers';
 
-import { ensure0x } from '@hyperlane-xyz/utils';
+import { Address, ensure0x } from '@hyperlane-xyz/utils';
+
+import { ContextSettings, KeyConfig } from '../context.js';
+
+import { impersonateAccount } from './fork.js';
+
+const ETHEREUM_ADDRESS_LENGTH = 42;
```
Comment (jmrossy) [inline typescript/cli/src/utils/keys.ts:10]: Please de-dupe with const in fork and use address utilities form `utils` package instead

File: typescript/cli/src/utils/keys.ts [hunk 31]
```diff
@@ -1,18 +1,109 @@
...
-export function assertSigner(signer: ethers.Signer) {
-  if (!signer || !ethers.Signer.isSigner(signer))
-    throw new Error('Signer is invalid');
+async function retrieveKey(
+  keyType: string,
+  keyConfig: KeyConfig,
+  skipConfirmation: boolean | undefined,
+): Promise<string> {
+  let key: string;
+  if (keyConfig.key) key = keyConfig.key;
```
Comment (jmrossy) [inline typescript/cli/src/utils/keys.ts:99]: We can just return directly instead of assigning `key`

File: typescript/cli/ci-test.sh [hunk 32]
```diff
@@ -2,191 +2,260 @@
...
+        chmod -R 777 /tmp/relayer /tmp/$CHAIN
+    done
+
+    anvil --chain-id 31337 -p 8545 --state /tmp/anvil1/state --gas-price 1 > /dev/null &
+    anvil --chain-id 31338 -p 8555 --state /tmp/anvil2/state --gas-price 1 > /dev/null &
+    sleep 1
+
+    set -e
+
+    echo "{}" > /tmp/empty-artifacts.json
```
Comment (jmrossy) [inline typescript/cli/ci-test.sh:64]: This is more related to the command prep than the anvil prep but nbd

File: typescript/cli/ci-test.sh [hunk 33]
```diff
@@ -2,191 +2,260 @@
...
+    pkill -f anvil
+}
+
+run_hyperlane_deploy_core_dry_run() {
+    BEFORE_DRY_RUN=$(cast balance $DEPLOYER --rpc-url http://localhost:8545)
+
+    echo "Dry-running contract deployments to Alfajores"
+    yarn workspace @hyperlane-xyz/cli run hyperlane deploy core --dry-run \
+        --targets alfajores \
+        --chains ./examples/anvil-chains.yaml \
```
Comment (jmrossy) [inline typescript/cli/ci-test.sh:81]: Nbd but you may not need this flag for this command

File: typescript/cli/ci-test.sh [hunk 34]
```diff
@@ -2,191 +2,260 @@
...
+    DEPLOYER=$(cast rpc eth_accounts | jq -r '.[0]');
+
+    run_hyperlane_deploy_core_dry_run;
+
+    reset_anvil;
+
+    run_hyperlane_deploy_core;
+    run_hyperlane_deploy_warp;
+    run_hyperlane_send_message;
+    run_hyperlane_send_transfer;
```
Comment (jmrossy) [inline typescript/cli/ci-test.sh:26]: Nice

**Review by jmrossy — COMMENTED** [review]
Mostly looks good! Thanks especially for the ci-test refactor and the CLI utilities improvements

File: typescript/cli/src/context.ts (hunk 28 above)
Comment (nbayindirli) [reply typescript/cli/src/context.ts:173]: The first provided chain will be forked if there are multiple.

This feature's next iteration will include multi-chain support and avoid naively choosing the first selection

File: typescript/cli/src/utils/fork.ts (hunk 29 above)
Comment (nbayindirli) [reply typescript/cli/src/utils/fork.ts:93]: There were one or two other steps here that I removed throughout development when migrating over from `keys.ts`, so having this trimmed did end up being useless. Thanks for the pointer on using `isValidAddressEvm`, this is much cleaner.

Updated.

File: typescript/cli/ci-test.sh (hunk 32 above)
Comment (nbayindirli) [reply typescript/cli/ci-test.sh]: ci-test was botched during rebase with `main` so will revisit this

File: typescript/cli/ci-test.sh (hunk 33 above)
Comment (nbayindirli) [reply typescript/cli/ci-test.sh]: as above, ci-test was botched during rebase with main so will revisit this
